File: implementations/java/src/main/java/io/rcl/Parser.java
package io.rcl;

import java.util.ArrayList;
import java.util.List;

final class Parser {
  private final List<Token> t;
  private int p = 0;

  Parser(String src) { t = Lexer.lex(src); }

  DocumentNode parse() {
    DocumentNode doc = new DocumentNode();
    while (!is(TokenKind.EOF)) doc.blocks.add(parseBlock());
    return doc;
  }

  private BlockNode parseBlock() {
    Token n = eat(TokenKind.IDENT, "unexpected token");
    String arg = null;
    if (is(TokenKind.STRING)) arg = eat(TokenKind.STRING, "unexpected token").text;
    eat(TokenKind.DO, "unexpected token");
    BlockNode b = new BlockNode(n.text, arg);
    List<String> seen = new ArrayList<>();
    while (!is(TokenKind.END)) {
      if (is(TokenKind.EOF)) throw err("missing end");
      Token cur = eat(TokenKind.IDENT, "unexpected token");
      Token nxt = peek();
      if (nxt.kind == TokenKind.EQ || nxt.kind == TokenKind.DOT) {
        String key = parseKey(cur.text);
        validateKey(key, seen, cur);
        eat(TokenKind.EQ, "unexpected token");
        AstNode v = parseValue();
        b.properties.put(key, v);
        b.propertyOrder.add(key);
      } else if (nxt.kind == TokenKind.DO || nxt.kind == TokenKind.STRING) {
        p--;
        BlockNode c = parseBlock();
        if (c.argument == null) b.blocks.add(c); else b.namedBlocks.add(c);
      } else throw err("unexpected token");
    }
    eat(TokenKind.END, "unexpected token");
    return b;
  }

  private String parseKey(String first) {
    String key = first;
    while (is(TokenKind.DOT)) {
      eat(TokenKind.DOT, "unexpected token");
      key += "." + eat(TokenKind.IDENT, "unexpected token").text;
    }
    return key;
  }

  private AstNode parseValue() {
    if (is(TokenKind.STRING)) return new StringNode(eat(TokenKind.STRING, "unexpected token").text);
    if (is(TokenKind.NUMBER)) return new NumberNode(Double.parseDouble(eat(TokenKind.NUMBER, "unexpected token").text));
    if (is(TokenKind.IDENT)) {
      String v = eat(TokenKind.IDENT, "unexpected token").text;
      if (v.equals("true") || v.equals("false")) return new BooleanNode(v.equals("true"));
      throw err("invalid bare identifier value");
    }
    if (is(TokenKind.LBR)) return parseArray();
    throw err("unexpected token");
  }

  private AstNode parseArray() {
    eat(TokenKind.LBR, "unexpected token");
    ArrayNode arr = new ArrayNode();
    if (is(TokenKind.RBR)) { eat(TokenKind.RBR, "unexpected token"); return arr; }
    arr.elements.add(parseValue());
    while (is(TokenKind.COMMA)) {
      eat(TokenKind.COMMA, "unexpected token");
      if (is(TokenKind.RBR)) throw err("trailing comma in array");
      arr.elements.add(parseValue());
    }
    if (!is(TokenKind.RBR)) throw err("missing ]");
    eat(TokenKind.RBR, "unexpected token");
    return arr;
  }

  private void validateKey(String key, List<String> seen, Token t0) {
    for (String ex : seen) {
      if (ex.equals(key)) throw new ParseError("duplicate key", t0.line, t0.col);
      if (isPrefix(ex, key) || isPrefix(key, ex)) throw new ParseError("key-path prefix conflict", t0.line, t0.col);
    }
    seen.add(key);
  }

  private boolean isPrefix(String a, String b) { return a.length() < b.length() && b.startsWith(a + "."); }
  private boolean is(TokenKind k) { return t.get(p).kind == k; }
  private Token peek() { return t.get(Math.min(p, t.size() - 1)); }
  private Token eat(TokenKind k, String msg) { if (!is(k)) throw err(msg); return t.get(p++); }
  private ParseError err(String msg) { Token x = t.get(Math.min(p, t.size() - 1)); return new ParseError(msg, x.line, x.col); }
}

final class ParseError extends RuntimeException {
  final int line;
  final int column;
  ParseError(String m, int l, int c) { super("line " + l + ", column " + c + ": " + m); line = l; column = c; }
}

File: implementations/kotlin/src/main/kotlin/io/rcl/Parser.kt
package io.rcl

class Parser private constructor(private val lexer: Lexer) {
  private var current: Token = lexer.nextToken()

  companion object {
    fun parse(text: String): DocumentNode = Parser(Lexer(text)).parse()
  }

  fun parse(): DocumentNode {
    val blocks = mutableListOf<BlockNode>()
    while (current.type != TokenType.EOF) blocks += parseBlock()
    return DocumentNode(blocks = blocks)
  }

  private fun parseBlock(): BlockNode {
    val name = current.value
    eat(TokenType.Identifier)

    var argument: String? = null
    if (current.type == TokenType.String) { argument = current.value; eat(TokenType.String) }
    eat(TokenType.Do)

    val properties = linkedMapOf<String, AstNode>()
    val blocks = linkedMapOf<String, BlockNode>()
    val named = mutableListOf<BlockNode>()
    val seen = linkedSetOf<String>()

    while (current.type != TokenType.End) {
      if (current.type == TokenType.EOF) throw error("missing end")
      if (current.type != TokenType.Identifier) throw error("expected identifier")
      when (peekToken().type) {
        TokenType.Do, TokenType.String -> {
          val child = parseBlock()
          if (child.argument != null) {
            named += child
          } else blocks[child.name] = child
        }
        TokenType.Equal, TokenType.Dot -> {
          val key = parsePropertyKey()
          ensureKeyValid(key, seen)
          eat(TokenType.Equal)
          properties[key] = parseValue()
        }
        else -> throw error("invalid statement")
      }
    }

    eat(TokenType.End)
    return BlockNode(name = name, argument = argument, properties = properties, blocks = blocks, namedBlocks = named)
  }

  private fun parsePropertyKey(): String {
    var key = current.value
    eat(TokenType.Identifier)
    while (current.type == TokenType.Dot) {
      eat(TokenType.Dot)
      if (current.type != TokenType.Identifier) throw error("expected identifier after dot")
      key += ".${current.value}"
      eat(TokenType.Identifier)
    }
    return key
  }

  private fun parseValue(): AstNode = when (current.type) {
    TokenType.String -> StringNode(value = current.value).also { eat(TokenType.String) }
    TokenType.Number -> NumberNode(value = current.value.toDouble()).also { eat(TokenType.Number) }
    TokenType.Identifier -> {
      val v = current.value
      eat(TokenType.Identifier)
      when (v) {
        "true" -> BooleanNode(value = true)
        "false" -> BooleanNode(value = false)
        else -> throw error("invalid bare value")
      }
    }
    TokenType.LBracket -> parseArray()
    else -> throw error("unexpected value")
  }

  private fun parseArray(): AstNode {
    eat(TokenType.LBracket)
    val elements = mutableListOf<AstNode>()
    if (current.type != TokenType.RBracket) {
      elements += parseValue()
      while (current.type == TokenType.Comma) {
        eat(TokenType.Comma)
        if (current.type == TokenType.RBracket) throw error("trailing comma in array")
        elements += parseValue()
      }
    }
    eat(TokenType.RBracket)
    return ArrayNode(elements = elements)
  }

  private fun eat(type: TokenType) {
    if (current.type != type) throw error("expected $type, got ${current.type}")
    current = lexer.nextToken()
  }

  private fun peekToken(): Token {
    val st = lexer.snapshot()
    val tok = lexer.nextToken()
    lexer.restore(st)
    return tok
  }

  private fun error(message: String): ParseError = ParseError(message, current.line, current.column)

  private fun ensureKeyValid(key: String, seen: MutableSet<String>) {
    if (seen.contains(key)) throw error("duplicate key")
    val parts = key.split(".")
    seen.forEach { existing ->
      val ex = existing.split(".")
      if (isPrefix(parts, ex) || isPrefix(ex, parts)) throw error("key conflict")
    }
    seen.add(key)
  }

  private fun isPrefix(left: List<String>, right: List<String>): Boolean {
    if (left.size >= right.size) return false
    left.indices.forEach { if (left[it] != right[it]) return false }
    return true
  }
}
